---
title: "Filtering and QC"
output:
  html_document: default
  pdf_document: default
---

__Author:__ Amanda Everitt  
__Began:__ 8/20/2018  
__Finished:__: 8/22/2018  


###[Packages Used]
- [Seurat](https://satijalab.org/seurat/)
- [scater](https://bioconductor.org/packages/release/bioc/html/scater.html)

###[Results]
Genes                
--------   --------------   ----------------
Initial        27998               
              -13065            Removed genes that didn't occur in >17 cells (0.1% of population)
Final          14933
---------  --------------   ----------------


Cells                
--------   --------------   ----------------
Initial        17823               
                -163            Removed cells with nUMI <500 or >100000
                -182            Removed cells with nGene <500 or >3000
                -82             Removed cells with a mitochondrial percent >30%
Final          17396
---------  --------------   ----------------


```{r "setup", include=FALSE}
rm(list=ls())
suppressPackageStartupMessages({
    library(knitr)
    library(Seurat)
    library(scater)
    library(SingleCellExperiment)
    library(mvoutlier)
    library(gridExtra)
})

out_dir = "outputs/output_01"
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
```

#1. Filter Data
###Load data 
```{r}
dataset_loc <- "/Users/AEveritt/projects/scRNAseq_L5_Tbr1/raw_data/mm10/"
ids <- c("L5HET/", "L5NULL/","L5WT/")

d10x.data <- sapply(ids, function(i){
  d10x <- Seurat::Read10X(file.path(dataset_loc,i))
  colnames(d10x) <- paste(sapply(strsplit(colnames(d10x),split="-"),'[[',1L),i,sep="-")
  d10x
})

experiment.data <- do.call("cbind", d10x.data)

experiment.aggregate <- CreateSeuratObject(
  experiment.data,
  project = "Siavash scRNA", 
  names.field = 2,
  min.cells = 17,
  names.delim = "\\-")
experiment.aggregate 
```

###Calculate proportion of mitochondrial, ribosomal, and pseudo genes in each cell
```{r}
mito.genes <- grep("^mt-", rownames(experiment.aggregate@data), value = T)
percent.mito <- Matrix::colSums(experiment.aggregate@raw.data[mito.genes, ]) / Matrix::colSums(experiment.aggregate@raw.data)
experiment.aggregate <- AddMetaData(
  object = experiment.aggregate,
  metadata = percent.mito,
  col.name= "percent.mito")

ribo.genes <- grep("^Rps|^Rpl|^Mrps|^Mrpl|^Gm", rownames(experiment.aggregate@data), value = T)
percent.ribo <- Matrix::colSums(experiment.aggregate@raw.data[ribo.genes, ]) / Matrix::colSums(experiment.aggregate@raw.data)
experiment.aggregate <- AddMetaData(
  object = experiment.aggregate,
  metadata = percent.ribo,
  col.name= "percent.ribo")

pseudo.genes <- grep("^Gm", rownames(experiment.aggregate@data), value = T)
percent.pseudo <- Matrix::colSums(experiment.aggregate@raw.data[pseudo.genes, ]) / Matrix::colSums(experiment.aggregate@raw.data)
experiment.aggregate <- AddMetaData(
  object = experiment.aggregate,
  metadata = percent.pseudo,
  col.name= "percent.pseudo")
```

###Filter based on nUMI, nGene, %mito (really fraction)
```{r}
p1a <- ggplot(experiment.aggregate@meta.data, aes(x=nUMI)) + 
  geom_histogram(binwidth=50, color="black", fill="white") + ggtitle("Distribution of nUMI") +
  geom_vline(aes(xintercept=500),color="red") + geom_vline(aes(xintercept=10000),color="red") + theme_bw()
p1b <- VlnPlot(experiment.aggregate, "nUMI", do.return = TRUE) + geom_hline(yintercept = 10000, color= "red")
p1b <- p1b + ggtitle("Distribution of nUMI by genotype") + xlab("Genotype") + ylab("Count") + theme_bw() + 
  theme(plot.title = element_text(color="black", size=12),
        axis.title.x = element_text(color="black", size=12),
        axis.title.y = element_text(color="black", size=12),
        legend.position = "none")

p2a <- ggplot(experiment.aggregate@meta.data, aes(x=nGene)) + 
  geom_histogram(binwidth=50, color="black", fill="white") + ggtitle("Distribution of nGene") +
  geom_vline(aes(xintercept=500),color="red") + geom_vline(aes(xintercept=3000),color="red") + theme_bw()
p2b <- VlnPlot(experiment.aggregate, "nGene", do.return = TRUE) + geom_hline(yintercept = 3000, color= "red")
p2b <- p2b + ggtitle("Distribution of nGene by genotype") + xlab("Genotype") + ylab("Count") + theme_bw() + 
  theme(plot.title = element_text(color="black", size=12),
        axis.title.x = element_text(color="black", size=12),
        axis.title.y = element_text(color="black", size=12),
        legend.position = "none")

p3a <- ggplot(experiment.aggregate@meta.data, aes(x=percent.mito)) + 
  geom_histogram(binwidth=0.01, color="black", fill="white") + ggtitle("Distribution of mitochondrial percentage") +
  geom_vline(aes(xintercept=0.3),color="red") + theme_bw()
p3b <- VlnPlot(experiment.aggregate, "percent.mito", do.return = TRUE) + geom_hline(yintercept = 0.3, color= "red")
p3b <- p3b + ggtitle("Distribution of mito % by genotype") + xlab("Genotype") + ylab("Count") + theme_bw() + 
  theme(plot.title = element_text(color="black", size=12),
        axis.title.x = element_text(color="black", size=12),
        axis.title.y = element_text(color="black", size=12),
        legend.position = "none")

grid.arrange(p1a,p2a, p3a, p1b, p2b, p3b, ncol=3)
```

```{r}
experiment.aggregate <- FilterCells(object = experiment.aggregate, subset.names = "nUMI", 
                                    low.thresholds = 500 , high.thresholds = 10000)
experiment.aggregate <- FilterCells(object = experiment.aggregate, subset.names = "nGene", 
                                    low.thresholds = 500 , high.thresholds = 3000)
experiment.aggregate <- FilterCells(object = experiment.aggregate, subset.names = "percent.mito", 
                                    low.thresholds = -Inf , high.thresholds = 0.3)
```

#####Note: ribo and pseudo genes had a seemingly normal distribution so they were not used for filtering (among other biological reasons)
```{r, results="hide"}
#Save Data 
save(experiment.aggregate, file=paste0(out_dir,"/experiment.aggregate.filtered.Rdata"))

#Clean up workspace
#detach("package:Seurat", unload=TRUE)
#detach("package:cowplot", unload=TRUE)
#gc()
#R.utils::gcDLLs() 
```





#2. Detect any outliers based on QC metrics
```{r, results="hide"}
core = SingleCellExperiment(assays = list(counts = as.matrix(experiment.aggregate@data)),
                            colData = as.data.frame(experiment.aggregate@meta.data), 
                            rowData = rownames(as.matrix(experiment.aggregate@data)))
core <- calculateQCMetrics(core)
core <- normalize(core)
```
```{r}
scater::plotExprsFreqVsMean(core)
```

###PCA based on expression 
```{r}
pca_exp <- scater::runPCA(core, use_coldata = F, detect_outliers = TRUE)
plotReducedDim(pca_exp, use_dimred="PCA", colour_by="orig.ident")
summary(pca_exp$outlier)
```

###PCA based on metadata 
```{r}
pca_coldata <- scater::runPCA(core, use_coldata = T,  detect_outliers = TRUE,
                              selected_variables = c("nGene", "pct_counts_in_top_500_features", "percent.mito","nUMI"))
summary(pca_coldata$outlier)
plotReducedDim(pca_coldata, use_dimred="PCA_coldata", colour_by="orig.ident")
plotReducedDim(pca_coldata, use_dimred="PCA_coldata", colour_by="outlier")

possible_outliers <- rownames(pca_coldata@colData[pca_coldata$outlier == T,])
save(possible_outliers, file=paste0(out_dir,"/possible_outliers.Rdata"))
```

```{r, out.width='100%'}
#scater::plotExplanatoryVariables(core)  #takes too long to plot for knitr and im impatient
knitr::include_graphics(paste0(out_dir,'/expl_plot.jpeg'))
```


```{r}
sessionInfo()
```

